java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -53,6 +53,110 @@ class ExpressionEvaluator(
         currentScope
 
     override fun <T> visit(expr: Expression.ObjectLiteralExpression<T>): Any = expr.value as Any
+
+    override fun <E, T> visit(expr: Expression.FromExpression<E, T>): Any {
+        var sequence = requireNotNull(currentScope.lookup<Any>(expr.source)) {
+            "${expr.source} is null in current scope."
+        }
+
+        if (sequence is List<*>) {
+            sequence = sequence.asSequence()
+        }
+
+        require(sequence is Sequence<*>) {
+            "${expr.source} of type ${sequence::class} cannot be converted to a Sequence"
+        }
+
+        return sequence.map { value ->
+            currentScope.set(expr.iterationVar, value as Any)
+            value
+        }
+    }
+
+    override fun <T> visit(expr: Expression.WhereExpression<T>): Any {
+        return (expr.qualifier.accept(this) as Sequence<T>).filter {
+            expr.expr.accept(this) == true
```

Comment:
I was baffled by you not using `it` in the filter statement. That's because the expression evaluates in the scope, which is different every time, right?

Another potential reason to avoid mutable scopes - maybe someday we could parallelize operations like `where` and `select`... though I don't think we should worry about this now...

Reply:
it's essentially doing. collection.filter { scope.lookup("p") < 5 }, because the Sequence created by the qualifier (in this case, the FROM expression) is setting "p", it doesn't refer to it. This is a consequence of linq, e.g. you could do

from p in collection
where random() < foo

so 'it' isn't involved, but may it, in which case, it appears as 'p' in the scope already